doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+  export A1
+}
+```
+
+Currently, this prevents us from using `A1` without a namespace, since we can't write an export for `import A(N=1).*`. @konnov suggested the introduction of `export *` for this end, which is still in consideration.
+
+#### `importedFrom`
+
+This is a simple reference to the import/export/instance statement that created a definition, also gathered during name collection. Used in flattening to distinguish definitions that come from instances from definitions that come from imports/exports.
+
+### Instance Flattener
+
+The instance flattener gets rid of instances, which is the most complicated part of flattening. We cannot use the same id for the same definition of different instances, as they might assume different values[^1].
+
+Each instance statement is replaced with an import statement and generates a new module. The new module has the constants replaced with `pure val` definitions (from the overrides in the instance statement), and all its names are fully-qualified, constructed with the previously defined namespace algorithm. The new import statement, replacing the instance statement, is a unqualified import of the new module.
+
+The names in the module containing the instance statement also have to be updated with a namespace, since they might refer to definitions that now include a namespace. We look at each name in the module and, if it refers to a definition that comes from an instance, (that is, its `importFrom` field has an instance), then we replace the name with a namespaced version of it (using the `namespaces` field). For this to work properly, any existing `import` and `export` statements should have been flattened before the instance flattener is run, ensuring that the definitions referred to by these names already have their proper namespaces.
+
+All of this tinkering on names and the addition of a matching import statement makes it so the module outputted by the instance flattener can be name resolved. The name resolution of this module will produce a lookup table where the names referring to instances point to the new modules (produced for each instance), which are no longer instances. Therefore, after instance flattening, we have instance-less modules and their corresponding lookup table.
+
+#### Example
+
+Take this module as example, where `A` is instanced twice with the same name:
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B {
+  import A(N=1) as A1
+  val b = A1::a // id 3
+}
+
+module C {
+  import A(N=0) as A1
+  val C = A1::a // id 4
+}
+```
+
+The instance flattener will create two new modules: `B::A1` and `C::A1`, with new ids for the definitions inside it.
+```bluespec
+module A {
+  const N: int // id 1
+  val a = N // id 2
+}
+
+module B::A1 {
+  pure val B::A1::N = 1 // id 5
+  pure val B::A1::a = B::A1::N // id 6
+}
+
+module C::A1 {
+  pure val C::A1::N = 0 // id 7
+  pure val C::A1::a = C::A1::N // id 8
+}
+
+module B {
+  import B::A1.*
```

Comment:
This is `import A(N=1) as A1` above, why does it become a `.*`?

Reply:
instances always become `.*` because we already take care of the namespaces inside the new module. My example in https://github.com/informalsystems/quint/pull/1087#discussion_r1284418538 might make this clearer.

Reply:
This is a consequence of the fact that we fully qualify the names inside `B::A1`. Since the names are already qualified inside the module, we can always import with `.*`.